api/v1alpha1/proxmoxcluster_types.go

```diff
@@ -50,25 +48,49 @@ type ProxmoxClusterSpec struct {
 	// +optional
 	SchedulerHints *SchedulerHints `json:"schedulerHints,omitempty"`
 
+	ClusterNetworkConfig `json:",inline"`
+}
+
+// ClusterNetworkConfig represents information about the cluster network configuration.
+type ClusterNetworkConfig struct {
 	// IPv4Config contains information about available IPV4 address pools and the gateway.
-	// this can be combined with ipv6Config in order to enable dual stack.
+	// this can be combined with ipv6Config to enable dual stack.
 	// either IPv4Config or IPv6Config must be provided.
 	// +optional
-	// +kubebuilder:validation:XValidation:rule="self.addresses.size() > 0",message="IPv4Config addresses must be provided"
-	IPv4Config *ipamicv1.InClusterIPPoolSpec `json:"ipv4Config,omitempty"`
+	IPv4Config *IPConfig `json:"ipv4Config,omitempty"`
 
 	// IPv6Config contains information about available IPV6 address pools and the gateway.
-	// this can be combined with ipv4Config in order to enable dual stack.
+	// this can be combined with ipv4Config to enable dual stack.
 	// either IPv4Config or IPv6Config must be provided.
 	// +optional
-	// +kubebuilder:validation:XValidation:rule="self.addresses.size() > 0",message="IPv6Config addresses must be provided"
-	IPv6Config *ipamicv1.InClusterIPPoolSpec `json:"ipv6Config,omitempty"`
+	IPv6Config *IPConfig `json:"ipv6Config,omitempty"`
 
 	// DNSServers contains information about nameservers used by machines network-config.
 	// +kubebuilder:validation:MinItems=1
 	DNSServers []string `json:"dnsServers"`
 }
 
+// IPConfig contains information about available IP config.
+type IPConfig struct {
+	// Addresses is a list of IP addresses that can be assigned. This set of
+	// addresses can be non-contiguous.
+	// +kubebuilder:validation:MinItems=1
+	Addresses []string `json:"addresses,omitempty"`
+
+	// Prefix is the network prefix to use.
+	// +kubebuilder:validation:Maximum=128
+	// +optional
+	Prefix int `json:"prefix,omitempty"`
+
+	// Gateway
+	// +optional
+	Gateway string `json:"gateway,omitempty"`
+
+	// DHCP indicates if DHCP should be used to assign IP addresses.
+	// +optional
+	DHCP bool `json:"dhcp,omitempty"`
```

Comment:
This needs to be a pointer if `false` is supposed to be a valid value

Reply:
but, this needs to be by default `false`

Reply:
You don't have to set a default. You can just treat nil like false

Reply:
so, i will have to change this fn and will be more complicated

```
// DHCPEnabled returns whether DHCP is enabled.
func (c ClusterNetworkConfig) DHCPEnabled() bool {
	switch {
	case (c.IPv6Config != nil && c.IPv6Config.DHCP) && (c.IPv4Config != nil && c.IPv4Config.DHCP):
		return true
	case (c.IPv6Config != nil && c.IPv6Config.DHCP) && c.IPv4Config == nil:
		return true
	case (c.IPv4Config != nil && c.IPv4Config.DHCP) && c.IPv6Config == nil:
		return true
	default:
		return false
	}
}

```

Reply:
Use `ptr.Deref` from `k8s.io/utils`

```go
// DHCPEnabled returns whether DHCP is enabled.
func (c ClusterNetworkConfig) DHCPEnabled() bool {
	switch {
	case (c.IPv6Config != nil && ptr.Deref(c.IPv6Config.DHCP, false)) && (c.IPv4Config != nil && ptr.Deref(c.IPv4Config.DHCP, false)):
		return true
	case (c.IPv6Config != nil && ptr.Deref(c.IPv6Config.DHCP, false)) && c.IPv4Config == nil:
		return true
	case (c.IPv4Config != nil && ptr.Deref(c.IPv4Config.DHCP, false)) && c.IPv6Config == nil:
		return true
	default:
		return false
	}
}
```

The bool pointer makes the DHCP setting clearer from the user's perspective because the explicit `false` stays `false`.

Reply:
Actually, to reduce dependencies you can just copy the ptr.Deref function into the api package. Keeps imports clean

Reply:
yes, you can check